### PR TITLE
[Flink-1815] [gelly] Add methods to read and write a Graph as adjacency list

### PR DESCRIPTION
**Pull request by fobeligi**
Added functionality for Jira Issue [FLINK-1815] Add methods to read and write a Graph as adjacency list

### CODE REVIEW

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 1]
```diff
@@ -408,6 +408,79 @@ public static GraphCsvReader fromCsvReader(String edgesPath, ExecutionEnvironmen
...
+	 *                   VERTEX_VALUE_DELIMITER: separating the source vertex-id from the vertex value, as well as the
+	 *                   target vertex-ids from the edge value.
+	 */
+	public void writeAsAdjacencyList(String filePath, String... delimiters) {
+
+		final String NEIGHBOR_DELIMITER = delimiters.length > 0 ? delimiters[0] : "\t";
+
+		final String VERTICES_DELIMITER = delimiters.length > 1 ? delimiters[1] : ",";
+
+		final String VERTEX_VALUE_DELIMITER = delimiters.length > 1 ? delimiters[2] : "-";
```
Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: Test length against "2".

File: flink-libraries/flink-gelly-scala/src/main/scala/org/apache/flink/graph/scala/Graph.scala [hunk 2]
```diff
@@ -1127,8 +1194,7 @@ TypeInformation : ClassTag](jgraph: jg.Graph[K, VV, EV]) {
    *
    * @param analytic the analytic to run on the Graph
    */
-  def run[T: TypeInformation : ClassTag](analytic: GraphAnalytic[K, VV, EV, T]):
-  GraphAnalytic[K, VV, EV, T] = {
+  def run[T: TypeInformation : ClassTag](analytic: GraphAnalytic[K, VV, EV, T])= {
```
Comment (greghogan) [inline flink-libraries/flink-gelly-scala/src/main/scala/org/apache/flink/graph/scala/Graph.scala]: Was this change intended?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 3]
```diff
@@ -408,6 +408,79 @@ public static GraphCsvReader fromCsvReader(String edgesPath, ExecutionEnvironmen
...
+
+
+		DataSet<Tuple2<K, VV>> vertices = this.getVerticesAsTuple2();
+
+		DataSet<Tuple3<K, K, EV>> edgesNValues = this.getEdgesAsTuple3();
```
Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: Do we need to convert the vertex and edge sets to tuples?

File: flink-libraries/flink-gelly-scala/src/main/scala/org/apache/flink/graph/scala/Graph.scala (hunk 2 above)
Comment (fobeligi) [inline flink-libraries/flink-gelly-scala/src/main/scala/org/apache/flink/graph/scala/Graph.scala]: No, I will revert the change.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java (hunk 3 above)
Comment (fobeligi) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: As I see now, we don't have to convert the vertex set to tuple2 set, so I already changed that.

Regarding the edges dataset, in order to write the Adjacency List file, I use the coGroup transformation to the Vertex dataset and EdgesAsTuple3 dataset, where the vertexId equals the source of the edge. 

In that case, even when a Vertex is source to no edges (e.g. has only incoming edges), I can still have the vertexId in the "coGrouped" dataset (I couldn't do that with a join).

I can't think how I could use the Edge dataset in a coGroup or similar transformation. 
Please let me know if you have any suggestions.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java (hunk 1 above)
Comment (fobeligi) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: You mean the error in this declaration: 
```java
final String VERTEX_VALUE_DELIMITER = delimiters.length > 1 ? delimiters[2] : "-";
```
and not to check directly for length greater than two, because in that way the user will have to provide all three delimiters or none.

**Comment by aljoscha** [comment]
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.
